controller/src/main/java/io/pravega/controller/server/retention/AutoRetentionBucketService.java

```diff
@@ -0,0 +1,160 @@
+    }
+
+    @Override
+    protected void doStart() {
+        RetryHelper.withIndefiniteRetriesAsync(() -> streamMetadataStore.getStreamsForBucket(bucketId, executor)
+                .thenAccept(streams -> retentionFutureMap.putAll(streams.stream()
+                        .map(s -> {
+                            String[] splits = s.split("/");
+                            return new StreamImpl(splits[0], splits[1]);
+                        })
+                        .collect(Collectors.toMap(s -> s, this::getStreamRetentionFuture))
+                )),
+                e -> log.warn("exception thrown getting streams for bucket {}, e = {}", bucketId, e), executor)
+                .thenAccept(x -> {
+                    log.info("streams collected for the bucket {}, registering for change notification and starting loop for processing notifications", bucketId);
+                    streamMetadataStore.registerBucketChangeListener(bucketId, this);
+                })
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStarted();
+                        notificationLoop = Futures.loop(this::isRunning, this::processNotification, executor);
+                    }
+                });
+    }
+
+    private CompletableFuture<Void> processNotification() {
+        return CompletableFuture.runAsync(() -> {
+            StreamNotification notification = Exceptions.handleInterrupted(() -> notifications.poll(1, TimeUnit.SECONDS));
+            if (notification != null) {
+                final StreamImpl stream;
+                switch (notification.getType()) {
+                    case StreamAdded:
+                        stream = new StreamImpl(notification.getScope(), notification.getStream());
+                        retentionFutureMap.putIfAbsent(stream, getStreamRetentionFuture(stream));
+                        break;
+                    case StreamRemoved:
+                        stream = new StreamImpl(notification.getScope(), notification.getStream());
+                        retentionFutureMap.remove(stream).cancel(true);
+                        break;
+                    case StreamUpdated:
+                        // if retention has been disabled then cancel the future.
+                        // if retention has been enabled then add the future if it does not exist.
+                        // For now we will do nothing.
+                        break;
+                    case ConnectivityError:
+                        log.info("Retention.StreamNotification for connectivity error");
+                        break;
+                }
+            }
+        }, executor);
+    }
+
+    private CompletableFuture<Void> getStreamRetentionFuture(StreamImpl stream) {
+        return RetryHelper.loopWithDelay(this::isRunning, () -> performRetention(stream),
```

Comment:
I don't immediately have any issue with this loop with delay, but I'm wondering about the overhead of maintaining such a look for each stream.

Reply:
Never mind, I can see in `ZKStream` that this is creating znodes under `retention/`.

Reply:
Well we have to maintain loops of O(#streams). Thats why we have bucketed them in buckets so that load can be distributed across controllers.  

As an optimization later on we could identify silent empty streams, streams which due to lack of activity have no new data coming in and truncation has purged all that was there in the stream (essentially no data is left in them) and purge them from this list.

Reply:
I'm wondering more about the scheduling than the memory footprint. How does the number of threads we need to handle all these loops grow with the number of streams?

Understood that bucketizing is to distributed the load.

Reply:
well all processing is asynchronous, and can even run on a single thread (though with large number of streams we may have significant delays). It is important to note though that with "loop" based computing we will only have exactly one scheduled/or running operation per stream. 
So its almost a fixed memory footprint which is O(#streams) because we have that many tasks posted into the executor's queue.

Reply:
Also, when controller starts and picks all existing streams, they are all scheduled at nearly same time with same frequencies.
(Any new stream, whenever added, will get its schedule posted then. )
So our work is not randomly distributed but most streams will start with their timers kicking in around similar times. But based on processing delays eventually they will get more scattered.

We can add some randomization for initial scheduling to ensure that workflows for all streams dont become eligible for execution concurrently.

Reply:
I just added the randomization in initial scheduling so that streams are more evenly distributed across RETENTION_FREQUENCY range.